storage/aptosdb/src/db_debugger/truncation/mod.rs

```diff
@@ -0,0 +1,414 @@
+            },
+            None => Ok(None),
+        }
+    }
+
+    fn root_exist_at_version(state_merkle_db: &DB, version: Version) -> Result<bool> {
+        Ok(state_merkle_db
+            .get::<JellyfishMerkleNodeSchema>(&NodeKey::new_empty_path(version))?
+            .is_some())
+    }
+
+    fn find_closest_node_version_at_or_before(
+        state_merkle_db: &DB,
+        version: Version,
+    ) -> Result<Option<Version>> {
+        let mut iter = state_merkle_db.rev_iter::<JellyfishMerkleNodeSchema>(Default::default())?;
+        iter.seek_for_prev(&NodeKey::new_empty_path(version))?;
+        Ok(iter.next().transpose()?.map(|item| item.0.version()))
+    }
+
+    fn truncate_ledger_db(
+        ledger_db: Arc<DB>,
+        current_version: Version,
+        target_version: Version,
+        batch_size: usize,
+    ) -> Result<()> {
+        let status = StatusLine::new(Progress::new(target_version));
+
+        let event_store = EventStore::new(Arc::clone(&ledger_db));
+        let transaction_store = TransactionStore::new(Arc::clone(&ledger_db));
+
+        let mut current_version = current_version;
+        while current_version > target_version {
+            let start_version =
+                std::cmp::max(current_version - batch_size as u64 + 1, target_version + 1);
+            let end_version = current_version + 1;
+            Self::truncate_ledger_db_single_batch(
+                &ledger_db,
+                &event_store,
+                &transaction_store,
+                start_version,
+                end_version,
+            )?;
+            current_version = start_version - 1;
+            status.set_current_version(current_version);
+        }
+        assert_eq!(current_version, target_version);
+        Ok(())
+    }
+
+    fn num_frozen_nodes_in_accumulator(num_leaves: u64) -> u64 {
+        2 * num_leaves - num_leaves.count_ones() as u64
+    }
+
+    fn truncate_transaction_accumulator(
+        ledger_db: &DB,
+        start_version: Version,
+        end_version: Version,
+        batch: &SchemaBatch,
+    ) -> Result<()> {
+        let num_frozen_nodes = Self::num_frozen_nodes_in_accumulator(end_version);
+        let mut iter = ledger_db.iter::<TransactionAccumulatorSchema>(ReadOptions::default())?;
+        iter.seek_to_last();
+        let (position, _) = iter.next().transpose()?.unwrap();
+        assert_eq!(position.to_postorder_index() + 1, num_frozen_nodes);
+
+        let num_frozen_nodes_after_this_batch =
+            Self::num_frozen_nodes_in_accumulator(start_version);
+
+        let mut num_nodes_to_delete = num_frozen_nodes - num_frozen_nodes_after_this_batch;
+
+        let start_position = Position::from_postorder_index(num_frozen_nodes_after_this_batch)?;
+        iter.seek(&start_position)?;
+
+        for item in iter {
+            let (position, _) = item?;
+            batch.delete::<TransactionAccumulatorSchema>(&position)?;
+            num_nodes_to_delete -= 1;
+        }
```

Comment:
why reading them out? you can calculate the keys, right?

Reply:
It's easy to read and understand. (~~I'm lazy~~)

Maybe I will change it to a different way (range delete + manual compaction?) for performance purpose.